src/Cardano/Transaction/Balancer/Collateral.purs

```diff
@@ -63,50 +63,89 @@ setTransactionCollateral
-    isSpendable = not <<< flip Set.member nonSpendableSet
-    coinsPerUtxoByte = pparams.coinsPerUtxoByte
-  collateral <- case mbCollateralUtxos of
-    -- if no collateral utxos are specified, use the wallet, but filter
-    -- the unspendable ones
-    Nothing -> do
-      let isSpendableUtxo = isSpendable <<< _.input <<< unwrap
-      (walletCollateral :: Array TransactionUnspentOutput) <- ExceptT $ liftAff
-        $
-          note CouldNotGetCollateral <$>
+  transaction
+  spendableUtxos = do
+  collateral <-
+    case (unwrap balancerConstraints).collateralUtxos of
+      -- if no collateral utxos are specified, use the wallet, but filter
+      -- the unspendable ones
+      Nothing -> do
+        (walletCollateral :: Array TransactionUnspentOutput) <-
+          ExceptT $ liftAff $ note CouldNotGetCollateral <$>
             getWalletCollateralAff
-      let
-        { yes: spendableUtxos, no: filteredUtxos } = Array.partition
-          isSpendableUtxo
-          walletCollateral
-      when (not $ Array.null filteredUtxos) do
-        logWithLevel Warn $ pprintTagSet
-          "Some of the collateral UTxOs returned by the wallet were marked as non-spendable and ignored"
-          (pprintUtxoMap (TransactionUnspentOutputs.toUtxoMap filteredUtxos))
-      pure spendableUtxos
-    -- otherwise, get all the utxos, filter out unspendable, and select
-    -- collateral using internal algo, that is also used in KeyWallet
-    Just utxoMap -> do
-      let
-        maxCollateralInputs = UInt.toInt pparams.maxCollateralInputs
-        mbCollateral =
-          Array.fromFoldable <$>
-            selectCollateral coinsPerUtxoByte maxCollateralInputs utxoMap
-      liftEither $ note (InsufficientCollateralUtxos utxoMap) mbCollateral
+        let
+          { yes: spendableCollUtxos, no: filteredUtxos } = Array.partition
+            isSpendable
+            walletCollateral
+        when (not $ Array.null filteredUtxos) do
+          logWithLevel Warn $ pprintTagSet
+            "Some of the collateral UTxOs returned by the wallet were marked as non-spendable and ignored"
+            (pprintUtxoMap $ toUtxoMap filteredUtxos)
+        let
+          collVal =
+            foldMap (Val.fromValue <<< _.amount <<< unwrap <<< _.output <<< unwrap)
+              spendableCollUtxos
+          minRequiredCollateral = BigNum.toBigInt $ unwrap Collateral.minRequiredCollateral
+        if (Val.getCoin collVal < minRequiredCollateral) then do
+          logWithLevel Warn $ pprintTagSet
+            "Filtered collateral UTxOs do not cover the minimum required \
+            \collateral, reselecting collateral using the internal CTL algorithm"
+            (pprintUtxoMap $ toUtxoMap spendableCollUtxos)
+          selectCollateralFromUtxos pparams $ Map.filter isPkhUtxo spendableUtxos
+        else pure spendableCollUtxos
+      -- otherwise, select collateral from the spendable utxos using
+      -- the KeyWallet algorithm
+      Just utxoMap -> do
```

Comment:
Don't we need a check here to ensure all these UTXOs are pkh UTXOs?

Reply:
We partition this set using `isSpendable` which includes the pkh check

Reply:
You're right